File: c/connect4/AlphabetaBot.cpp
#include "stdafx.h"
#include "AlphabetaBot.h"
#include "bits.h"
#include "evaluation.h"
#include <iostream>


AlphabetaBot::AlphabetaBot(std::uint32_t depth, bool pvsEnabled, TranspositionTable* table, bool verbose)
{
	this->depth = depth;
	this->table = table;
	this->pvsEnabled = pvsEnabled;
	this->verbose = verbose;
}


AlphabetaBot::~AlphabetaBot(void)
{
}

std::uint64_t AlphabetaBot::search(Connect4* state)
{
	this->nodes = 1;
	std::uint64_t bestMove;
	std::int32_t alpha = evaluation::LOSS_SCORE - this->depth;
	std::int32_t beta = -alpha;
	bool foundPV = false;
	std::uint64_t moves = state->tokenMoves();
	while(moves)
	{
		std::uint64_t move = bits::highestBit(moves);
		moves ^= move;
		state->tokenMove(move);
		std::int32_t score;
		if (this->pvsEnabled && foundPV)
		{
			score = -this->alphabeta(state, this->depth - 1, -alpha - 1, -alpha);
			if (alpha < score)
			{
				score = -this->alphabeta(state, this->depth - 1, -beta, -alpha);
			}
		}
		else
		{
			score = -this->alphabeta(state, this->depth - 1, -beta, -alpha);
		}
		state->tokenUndo(move);
		if(score > alpha)
		{
			alpha = score;
			bestMove = move;
			foundPV = true;
		}
	}
	if (this->verbose)
	{
		std::cout << "nodes searched: " << this->nodes << std::endl;
		if (this->table)
		{
			this->table->printStats();
		}
	}
	return bestMove;
}

std::int32_t AlphabetaBot::alphabeta(Connect4* state, std::uint32_t depth, std::int32_t alpha, std::int32_t beta)
{
	this->nodes += 1;
	if (state->opponentWon())
	{
		return evaluation::LOSS_SCORE - depth;
	}
	if (state->isBoardFull())
	{
		return 0;
	}
	if (depth <= 0)
	{
		return evaluation::evaluate(state);
	}

	if (this->table)
	{
		TranspositionEntry entry;
		if (this->table->load(state->id(), &entry) && entry.depth >= depth)
		{
			switch (entry.type)
			{
			case UPPER_BOUND:
				beta = entry.score;
				if (alpha >= beta)
				{
					return alpha;
				}
				break;
			case LOWER_BOUND:
				alpha = entry.score;
				if (alpha >= beta)
				{
					return beta;
				}
				break;
			case EXACT:
				return entry.score;
			}
		}
	}

	ScoreType type = UPPER_BOUND;
	bool foundPV = false;
	std::uint64_t moves = state->tokenMoves();
	while(moves)
	{
		std::uint64_t move = bits::highestBit(moves);
		moves ^= move;
		state->tokenMove(move);
		std::int32_t score;
		if (this->pvsEnabled && foundPV)
		{
			score = -this->alphabeta(state, depth - 1, -alpha - 1, -alpha);
			if (alpha < score)
			{
				score = -this->alphabeta(state, depth - 1, -beta, -alpha);
			}
		}
		else
		{
			score = -this->alphabeta(state, depth - 1, -beta, -alpha);
		}
		state->tokenUndo(move);
		if (score > alpha)
		{
			if (score >= beta)
			{
				alpha = beta;
				type = LOWER_BOUND;
				break;
			}
			alpha = score;
			foundPV = true;
			type = EXACT;
		}
	}

	if (this->table)
	{
		TranspositionEntry entry = TranspositionEntry();
		entry.id = state->id();
		entry.type = type;
		entry.score = alpha;
		entry.depth = depth;
		this->table->store(entry);
	}
	return alpha;
}

File: java/src/main/java/com/etherblood/connect4/AlphaBetaBot.java
package com.etherblood.connect4;

import com.etherblood.connect4.eval.Evaluation;
import com.etherblood.connect4.transpositions.TableEntry;
import com.etherblood.connect4.transpositions.TranspositionTable;

/**
 *
 * @author Philipp
 */
public class AlphaBetaBot {

    private final Connect4State state;
    private final Evaluation eval;
    private final TranspositionTable table;
    private final int depth;
    private final boolean pvsEnabled, verbose;
    private long searchedNodes = 0;

    public AlphaBetaBot(Connect4State state, Evaluation eval, TranspositionTable table, int depth, boolean pvsEnabled, boolean verbose) {
        this.state = state;
        this.eval = eval;
        this.table = table;
        this.depth = depth;
        this.pvsEnabled = pvsEnabled;
        this.verbose = verbose;
    }

    public long search() {
        long startNanos = System.nanoTime();
        searchedNodes = 1;
        long bestMove = 0;
        int alpha = eval.lossScore() - depth;
        int beta = -alpha;
        boolean foundPV = false;
        long moves = state.availableMoves();
        while (moves != 0) {
            long move = Long.lowestOneBit(moves);//TODO: move order
            moves ^= move;
            state.move(move);
            int score;
            if (pvsEnabled && foundPV) {
                score = -alphabeta(depth - 1, -alpha - 1, -alpha);
                if (alpha < score) {
                    score = -alphabeta(depth - 1, -beta, -alpha);
                }
            } else {
                score = -alphabeta(depth - 1, -beta, -alpha);
            }
            state.unmove(move);
            if (score > alpha) {
                alpha = score;
                bestMove = move;
                foundPV = true;
            }
        }
        if (verbose) {
            long endNanos = System.nanoTime();
            System.out.println("nodes searched: " + searchedNodes + " in " + (endNanos - startNanos) / 1_000_000 + "ms");
            System.out.println("score: " + (state.activePlayer() == 0 ? alpha : -alpha));
            if (table != null) {
                table.printStats();
            }
        }
        return bestMove;
    }

    public int alphabeta(int depth, int alpha, int beta) {
        searchedNodes++;
        if (state.opponentWon()) {
            return eval.lossScore() - depth;
        }
        if (state.isBoardFull()) {
            return eval.drawScore();
        }
        if (depth == 0) {
            return eval.evaluate();
        }
        TableEntry entry = null;
        if (table != null) {
            entry = table.load(state.id());
        }
        if (entry != null && entry.depth >= depth) {
            switch (entry.type) {
                case TableEntry.UPPER_BOUND:
                    beta = entry.score;
                    if (alpha >= beta) {
                        return alpha;
                    }
                    break;
                case TableEntry.LOWER_BOUND:
                    alpha = entry.score;
                    if (alpha >= beta) {
                        return beta;
                    }
                    break;
                case TableEntry.EXACT:
                    return entry.score;
                default:
                    throw new AssertionError(entry.type);

            }
        }
        int type = TableEntry.UPPER_BOUND;
        boolean foundPV = false;
        long moves = state.availableMoves();
        while (moves != 0) {
            long move = Long.lowestOneBit(moves);//TODO: move order
            moves ^= move;
            state.move(move);
            int score;
            if (pvsEnabled && foundPV) {
                score = -alphabeta(depth - 1, -alpha - 1, -alpha);
                if (alpha < score) {
                    score = -alphabeta(depth - 1, -beta, -alpha);
                }
            } else {
                score = -alphabeta(depth - 1, -beta, -alpha);
            }
            state.unmove(move);
            if (score > alpha) {
                if (score >= beta) {
                    alpha = beta;
                    type = TableEntry.LOWER_BOUND;
                    break;
                }
                alpha = score;
                foundPV = true;
                type = TableEntry.EXACT;
            }
        }
        if (table != null) {
            table.storeRaw(state.id(), type, alpha, depth, 1);
        }
        return alpha;
    }
}
